### ISSUE TITLE
shadow_unrelated false positive with question mark

### ISSUE THREAD

**Issue by camsteffen** [rust-lang/rust-clippy#6563]
Lint name: shadow_unrelated

I tried this code:

```rust
#[warn(clippy::shadow_unrelated)]
fn foo() -> Option<i32> {
    let a = 1_i32;
    let a = Some(1).map(|_| a)?;
    Some(a)
}
```

I expected to see this happen: no lint

Instead, this happened: shadow_unrelated lint

I needed a closure and a question mark to reproduce this. 🤷

**Comment by dpathakj** [comment]
This might be the same issue as https://github.com/rust-lang/rust-clippy/issues/6141.

**Comment by whatisaphone** [comment]
On the playground it repros on stable and nightly with just a closure.

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f2416025978fe7e5c9407e811a82a744)

```rust
#![deny(clippy::shadow_unrelated)]

fn main() {
    let foo = 0;
    let foo = || foo;
}
```

Expected result: no error, since `|| foo` refers to the variable `foo`


<details>
<summary>Actual result: shadow_unrelated error</summary>

```
error: `foo` is being shadowed
 --> src/main.rs:5:9
  |
5 |     let foo = || foo;
  |         ^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(clippy::shadow_unrelated)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^
note: initialization happens here
 --> src/main.rs:5:15
  |
5 |     let foo = || foo;
  |               ^^^^^^
note: previous binding is here
 --> src/main.rs:4:9
  |
4 |     let foo = 0;
  |         ^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#shadow_unrelated
```

</details>

**Comment by lopopolo** [comment]
This is another bizarre false positive I get:

```
warning: `self.state.as_deref_mut().ok_or_else(InterpreterExtractError::new)?` is being shadowed
  --> artichoke-backend/src/parser.rs:32:21
   |
32 |         let state = self.state.as_deref_mut().ok_or_else(InterpreterExtractError::new)?;
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> artichoke-backend/src/lib.rs:2:9
   |
2  | #![warn(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::shadow_unrelated)]` implied by `#[warn(clippy::pedantic)]`
note: initialization happens here
  --> artichoke-backend/src/parser.rs:32:21
   |
32 |         let state = self.state.as_deref_mut().ok_or_else(InterpreterExtractError::new)?;
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: previous binding is here
  --> artichoke-backend/src/parser.rs:31:36
   |
31 |     fn add_fetch_lineno(&mut self, val: usize) -> Result<usize, Self::Error> {
   |                                    ^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#shadow_unrelated
```
